### ISSUE TITLE
Repeated NullPointerExceptions encountered when Vaadin project uses push, and includes a custom widget set

### ISSUE THREAD

**Issue by nabil-hassan** [johndevs/gradle-vaadin-plugin#283]
Hi there,

Thanks for the great work on this plugin. 

We are using version 1.0, and encounter issues on our project, which has Vaadin push enabled, and a custom widgetset. 

The stack trace is shown below, and occurs when we try to run the project from the command line using 'vaadinRun' OR when attempting to use eclipse wtp to run/debug the project.

I have done some investigation, and the root cause is that when the project has a widget set to be compiled, your plugin includes gwt-dev-2.7.0.vaadin4 on the classpath. 

This dependency includes Jetty 8 Servlet3Continuation class, which means the AtmosphereFramework uses JettyServlet30AsyncSupportWithWebSocket, instead of JSR356AsyncSupport. 

See: org.atmosphere.cpr.DefaultAsyncSupportResolver.detectWebSocketPresent(boolean, boolean)

When this dependency is not included, Atmosphere uses the JSR356 support, and we have no issues.

Please can you look at this when you have time?

```
WARNING: StandardWrapperValve[SeveneyeUIServlet]: Servlet.service() for servlet SeveneyeUIServlet threw exception
java.lang.NullPointerException
at org.eclipse.jetty.websocket.WebSocketFactory.upgrade(WebSocketFactory.java:238)
at org.eclipse.jetty.websocket.WebSocketFactory.acceptWebSocket(WebSocketFactory.java:396)
at org.atmosphere.container.JettyWebSocketUtil.doService(JettyWebSocketUtil.java:60)
at org.atmosphere.container.JettyServlet30AsyncSupportWithWebSocket.service(JettyServlet30AsyncSupportWithWebSocket.java:52)
at org.atmosphere.cpr.AtmosphereFramework.doCometSupport(AtmosphereFramework.java:2075)
at com.vaadin.server.communication.PushRequestHandler.handleRequest(PushRequestHandler.java:248)
at com.vaadin.server.VaadinService.handleRequest(VaadinService.java:1422)
at com.vaadin.server.VaadinServlet.service(VaadinServlet.java:379)
at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
at org.apache.catalina.core.StandardWrapper.service(StandardWrapper.java:1682)
at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:344)
at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:214)
at org.glassfish.tyrus.servlet.TyrusServletFilter.doFilter(TyrusServletFilter.java:255)
at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:256)
at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:214)
at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:316)
at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:160)
at org.apache.catalina.core.StandardPipeline.doInvoke(StandardPipeline.java:734)
at org.apache.catalina.core.StandardPipeline.invoke(StandardPipeline.java:673)
at com.sun.enterprise.web.WebPipeline.invoke(WebPipeline.java:99)
at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:174)
at org.apache.catalina.connector.CoyoteAdapter.doService(CoyoteAdapter.java:415)
at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:282)
at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:459)
at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:167)
at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:206)
at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:180)
at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:235)
at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:119)
at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:283)
at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:200)
at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:132)
at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:111)
at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:77)
at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:536)
at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:112)
at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:117)
at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(WorkerThreadIOStrategy.java:56)
at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:137)
at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:591)
at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:571)
at java.lang.Thread.run(Thread.java:745)
```

**Comment by johndevs** [comment]
Hi,

The GWT jars should only be added the GWT compilation classpath and not to the WAR or to the classpath of vaadinRun if the plugin is managing the dependencies and you haven't added duplicate Vaadin dependencies (plugin adds the Vaadin dependencies so you shouldn't add them yourself). 

Also, ensure you don't have ``vaadin.manageDependencies = false``, if you do, you need to make sure to exclude those dependencies yourself.

If you could you post the output of running ``gradle dependencies`` and update to ``1.0.1`` it would help to resolve the issue further.

**Comment by vyazelenko** [comment]
We have the same issue when running application from within Eclipse. Here is minimal reproducer `build.gradle` file:
```
plugins {
  id "war"
  id "eclipse-wtp"
  id "fi.jasoft.plugin.vaadin" version "1.0.1"
}

repositories {
  jcenter()
}

vaadin {
  version '7.7.6'
  
  vaadinCompile {
    widgetset 'example.WidgetSet'
  } 
}
```
When I run `gradle dependencies` I get the following output:
```
$ gradle dependencies
:dependencies

------------------------------------------------------------
Root project
------------------------------------------------------------

archives - Configuration for archive artifacts.
No dependencies

compile - Dependencies for source set 'main'.
No dependencies

compileClasspath - Compile classpath for source set 'main'.
No dependencies

compileOnly - Compile dependencies for source set 'main'.
No dependencies

default - Configuration for default artifacts.
No dependencies

providedCompile - Additional compile classpath for libraries that should not be part of the WAR archive.
No dependencies

providedRuntime - Additional runtime classpath for libraries that should not be part of the WAR archive.
No dependencies

runtime - Runtime dependencies for source set 'main'.
No dependencies

testCompile - Dependencies for source set 'test'.
No dependencies

testCompileClasspath - Compile classpath for source set 'test'.
No dependencies

testCompileOnly - Compile dependencies for source set 'test'.
No dependencies

testRuntime - Runtime dependencies for source set 'test'.
No dependencies

vaadin-client - Libraries needed for compiling the widgetset.
+--- com.vaadin:vaadin-client:7.7.6
|    +--- com.vaadin:vaadin-shared:7.7.6
|    +--- com.vaadin:vaadin-server:7.7.6
|    |    +--- com.vaadin:vaadin-shared:7.7.6
|    |    \--- org.jsoup:jsoup:1.8.3
|    \--- com.vaadin.external.gwt:gwt-elemental:2.7.0.vaadin4
|         \--- com.vaadin.external.gwt:gwt-user:2.7.0.vaadin4
|              \--- javax.validation:validation-api:1.0.0.GA
+--- com.vaadin:vaadin-client-compiler:7.7.6
|    +--- com.vaadin:vaadin-client:7.7.6 (*)
|    +--- com.vaadin.external.gwt:gwt-dev:2.7.0.vaadin4
|    |    +--- org.ow2.asm:asm:5.0.3
|    |    +--- org.ow2.asm:asm-util:5.0.3
|    |    |    \--- org.ow2.asm:asm-tree:5.0.3
|    |    |         \--- org.ow2.asm:asm:5.0.3
|    |    \--- org.ow2.asm:asm-commons:5.0.3
|    |         \--- org.ow2.asm:asm-tree:5.0.3 (*)
|    \--- com.vaadin:vaadin-sass-compiler:0.9.13
|         +--- org.w3c.css:sac:1.3
|         +--- com.vaadin.external.flute:flute:1.3.0.gg2
|         \--- com.yahoo.platform.yui:yuicompressor:2.4.8
|              \--- rhino:js:1.7R2
\--- javax.validation:validation-api:1.0.0.GA

vaadin-javadoc - Libraries for compiling JavaDoc for a Vaadin project.
+--- javax.portlet:portlet-api:2.0
+--- javax.servlet:javax.servlet-api:3.1.0
\--- com.vaadin:vaadin-push:7.7.6
     \--- com.vaadin.external.atmosphere:atmosphere-runtime:2.2.9.vaadin2
          \--- com.vaadin.external.slf4j:vaadin-slf4j-jdk14:1.6.1

vaadin-push - Libraries needed for using Vaadin Push features.
No dependencies

vaadin-run-server - Libraries for running the embedded server
+--- fi.jasoft.plugin:gradle-vaadin-plugin:1.0.1
\--- fish.payara.extras:payara-embedded-web:4.1.153

vaadin-server - Libraries needed by Vaadin server side applications.
+--- com.vaadin:vaadin-server:7.7.6
|    +--- com.vaadin:vaadin-sass-compiler:0.9.13
|    |    +--- org.w3c.css:sac:1.3
|    |    \--- com.vaadin.external.flute:flute:1.3.0.gg2
|    +--- com.vaadin:vaadin-shared:7.7.6
|    \--- org.jsoup:jsoup:1.8.3
+--- com.vaadin:vaadin-themes:7.7.6
\--- javax.servlet:javax.servlet-api:3.1.0

vaadin-superdevmode - Libraries needed by Vaadin Superdevmode.
+--- org.eclipse.jetty.aggregate:jetty-all-server:8.1.15.v20140411
|    +--- org.eclipse.jetty.orbit:javax.servlet:3.0.0.v201112011016
|    +--- org.eclipse.jetty.orbit:javax.security.auth.message:1.0.0.v201108011116
|    +--- org.eclipse.jetty.orbit:javax.mail.glassfish:1.4.1.v201005082020
|    |    \--- org.eclipse.jetty.orbit:javax.activation:1.1.0.v201105071233
|    +--- org.eclipse.jetty.orbit:javax.activation:1.1.0.v201105071233
|    \--- org.eclipse.jetty.orbit:javax.annotation:1.1.0.v201108011116
+--- fi.jasoft.plugin:gradle-vaadin-plugin:1.0.1
|    \--- org.codehaus.groovy.modules.http-builder:http-builder:0.7.1
|         +--- org.apache.httpcomponents:httpclient:4.2.1
|         |    +--- org.apache.httpcomponents:httpcore:4.2.1
|         |    +--- commons-logging:commons-logging:1.1.1
|         |    \--- commons-codec:commons-codec:1.6
|         +--- net.sf.json-lib:json-lib:2.3
|         |    +--- commons-beanutils:commons-beanutils:1.8.0
|         |    |    \--- commons-logging:commons-logging:1.1.1
|         |    +--- commons-collections:commons-collections:3.2.1
|         |    +--- commons-lang:commons-lang:2.4
|         |    +--- commons-logging:commons-logging:1.1.1
|         |    \--- net.sf.ezmorph:ezmorph:1.0.6
|         |         \--- commons-lang:commons-lang:2.3 -> 2.4
|         +--- net.sourceforge.nekohtml:nekohtml:1.9.16
|         |    \--- xerces:xercesImpl:2.9.1
|         \--- xml-resolver:xml-resolver:1.2
+--- org.ow2.asm:asm:5.0.3
+--- org.ow2.asm:asm-commons:5.0.3
|    \--- org.ow2.asm:asm-tree:5.0.3
|         \--- org.ow2.asm:asm:5.0.3
\--- javax.servlet.jsp:jsp-api:2.2

vaadin-testbench - Libraries needed by Vaadin Testbench.
No dependencies

vaadin-theme-compiler - Libraries needed for SASS theme compilation
No dependencies

BUILD SUCCESSFUL

Total time: 0.749 secs
```
However when I import sample project into Eclipse I see the following entries in the classpath:
![image](https://cloud.githubusercontent.com/assets/696855/21502561/045094c0-cc51-11e6-9568-41ec0fe251af.png)

But if remove `vaadinCompile` section of the `vaadin` block I get the following results.
```
$ gradle dependencies
:dependencies

------------------------------------------------------------
Root project
------------------------------------------------------------

archives - Configuration for archive artifacts.
No dependencies

compile - Dependencies for source set 'main'.
No dependencies

compileClasspath - Compile classpath for source set 'main'.
No dependencies

compileOnly - Compile dependencies for source set 'main'.
No dependencies

default - Configuration for default artifacts.
No dependencies

providedCompile - Additional compile classpath for libraries that should not be part of the WAR archive.
No dependencies

providedRuntime - Additional runtime classpath for libraries that should not be part of the WAR archive.
No dependencies

runtime - Runtime dependencies for source set 'main'.
No dependencies

testCompile - Dependencies for source set 'test'.
No dependencies

testCompileClasspath - Compile classpath for source set 'test'.
No dependencies

testCompileOnly - Compile dependencies for source set 'test'.
No dependencies

testRuntime - Runtime dependencies for source set 'test'.
No dependencies

vaadin-client - Libraries needed for compiling the widgetset.
\--- com.vaadin:vaadin-client-compiled:7.7.6

vaadin-javadoc - Libraries for compiling JavaDoc for a Vaadin project.
+--- javax.portlet:portlet-api:2.0
+--- javax.servlet:javax.servlet-api:3.1.0
\--- com.vaadin:vaadin-push:7.7.6
     \--- com.vaadin.external.atmosphere:atmosphere-runtime:2.2.9.vaadin2
          \--- com.vaadin.external.slf4j:vaadin-slf4j-jdk14:1.6.1

vaadin-push - Libraries needed for using Vaadin Push features.
No dependencies

vaadin-run-server - Libraries for running the embedded server
+--- fi.jasoft.plugin:gradle-vaadin-plugin:1.0.1
\--- fish.payara.extras:payara-embedded-web:4.1.153

vaadin-server - Libraries needed by Vaadin server side applications.
+--- com.vaadin:vaadin-server:7.7.6
|    +--- com.vaadin:vaadin-sass-compiler:0.9.13
|    |    +--- org.w3c.css:sac:1.3
|    |    \--- com.vaadin.external.flute:flute:1.3.0.gg2
|    +--- com.vaadin:vaadin-shared:7.7.6
|    \--- org.jsoup:jsoup:1.8.3
+--- com.vaadin:vaadin-themes:7.7.6
\--- javax.servlet:javax.servlet-api:3.1.0

vaadin-superdevmode - Libraries needed by Vaadin Superdevmode.
+--- org.eclipse.jetty.aggregate:jetty-all-server:8.1.15.v20140411
|    +--- org.eclipse.jetty.orbit:javax.servlet:3.0.0.v201112011016
|    +--- org.eclipse.jetty.orbit:javax.security.auth.message:1.0.0.v201108011116
|    +--- org.eclipse.jetty.orbit:javax.mail.glassfish:1.4.1.v201005082020
|    |    \--- org.eclipse.jetty.orbit:javax.activation:1.1.0.v201105071233
|    +--- org.eclipse.jetty.orbit:javax.activation:1.1.0.v201105071233
|    \--- org.eclipse.jetty.orbit:javax.annotation:1.1.0.v201108011116
+--- fi.jasoft.plugin:gradle-vaadin-plugin:1.0.1
|    \--- org.codehaus.groovy.modules.http-builder:http-builder:0.7.1
|         +--- org.apache.httpcomponents:httpclient:4.2.1
|         |    +--- org.apache.httpcomponents:httpcore:4.2.1
|         |    +--- commons-logging:commons-logging:1.1.1
|         |    \--- commons-codec:commons-codec:1.6
|         +--- net.sf.json-lib:json-lib:2.3
|         |    +--- commons-beanutils:commons-beanutils:1.8.0
|         |    |    \--- commons-logging:commons-logging:1.1.1
|         |    +--- commons-collections:commons-collections:3.2.1
|         |    +--- commons-lang:commons-lang:2.4
|         |    +--- commons-logging:commons-logging:1.1.1
|         |    \--- net.sf.ezmorph:ezmorph:1.0.6
|         |         \--- commons-lang:commons-lang:2.3 -> 2.4
|         +--- net.sourceforge.nekohtml:nekohtml:1.9.16
|         |    \--- xerces:xercesImpl:2.9.1
|         \--- xml-resolver:xml-resolver:1.2
+--- org.ow2.asm:asm:5.0.3
+--- org.ow2.asm:asm-commons:5.0.3
|    \--- org.ow2.asm:asm-tree:5.0.3
|         \--- org.ow2.asm:asm:5.0.3
\--- javax.servlet.jsp:jsp-api:2.2

vaadin-testbench - Libraries needed by Vaadin Testbench.
No dependencies

vaadin-theme-compiler - Libraries needed for SASS theme compilation
No dependencies

BUILD SUCCESSFUL

Total time: 0.748 secs
```
And in the Eclipse I see this:
![image](https://cloud.githubusercontent.com/assets/696855/21502589/3189d848-cc51-11e6-8fad-d44323d6291e.png)

So `vaadinCompile` is the culprit of the problem. It seems that Buildship picks up dependencies from  `vaadinCompile` block when it is defined.

Environment:
- Eclipse
```
Eclipse Java EE IDE for Web Developers.

Version: Neon.2 Release (4.6.2)
Build id: 20161208-0600
```
-   Buildship
```
Eclipse Plug-ins for Gradle	1.0.21.v20161010-1640
```

**Comment by johndevs** [comment]
When you define a widgetset with the ``vaadinCompile.widgetset`` (or when you add an addon) the plugin will include different dependencies than when you don't have a widgetset. 

When you have a widgetset then the GWT dependencies along with other vaadin-compiler related dependencies will need to be included so GWT client side code in the project as well as in any addons can be compiled into a application widgetset.

If you don't have a widgetset, then ``vaadin-client-compiled`` will be included which contains a pre-compiled widgetset. This is called the DefaultWidgetset and is used by Vaadin by default when you don't have custom client side code.

What you see in Eclipse is correct, when you have a widgetset the GWT dependencies should be on your project classpath. Otherwise you wouldn't be able to write GWT client side code in your project. The fact that they are visible there does not mean that they will be included in the packaged WAR nor will they be in the WTP deployment assembly (this was fixed in #285). The whole ``vaadin-client`` configuration should be excluded from packaging in this case.

Please try the latest version of the plugin (1.1) where #285 is fixed and run ``cleanEclipse eclipse`` to get the WTP deployment correct. I think that will solve your issue.

**Comment by vyazelenko** [comment]
@johndevs Just tried with `1.1.1` and it seems to work fine.
